Software/src/LedDeviceUdp.cpp

```diff
@@ -0,0 +1,276 @@
+/*
+ * LedDeviceUdp.cpp
+ *
+ *	Created on: 06.09.2011
+ *		Author: Mike Shatohin (brunql), Timur Sattarov
+ *		Project: Lightpack
+ *
+ *	Lightpack is very simple implementation of the backlight for a laptop
+ *
+ *	Copyright (c) 2011 Mike Shatohin, mikeshatohin [at] gmail.com
+ *
+ *	Lightpack is free software: you can redistribute it and/or modify
+ *	it under the terms of the GNU General Public License as published by
+ *	the Free Software Foundation, either version 2 of the License, or
+ *	(at your option) any later version.
+ *
+ *	Lightpack is distributed in the hope that it will be useful,
+ *	but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	See the
+ *	GNU General Public License for more details.
+ *
+ *	You should have received a copy of the GNU General Public License
+ *	along with this program.	If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "LedDeviceUdp.hpp"
+#include "PrismatikMath.hpp"
+#include "Settings.hpp"
+#include "enums.hpp"
+#include "debug.h"
+
+using namespace SettingsScope;
+
+LedDeviceUdp::LedDeviceUdp(QObject * parent) : AbstractLedDevice(parent)
+{
+	DEBUG_LOW_LEVEL << Q_FUNC_INFO;
+
+	m_gamma = Settings::getDeviceGamma();
+	m_brightness = Settings::getDeviceBrightness();
+
+	m_MoteDevice = NULL;
+	m_writeBufferHeader.append((char)255);
+
+	// Open here otherwise config wizard doesn't work
+	open();
+}
+
+LedDeviceUdp::~LedDeviceUdp()
+{
+	close();
+}
+
+void LedDeviceUdp::close()
+{
+	if (m_MoteDevice != NULL) {
+		m_MoteDevice->close();
+
+		delete m_MoteDevice;
+		m_MoteDevice = NULL;
+	}
+}
+
+void LedDeviceUdp::setColors(const QList<QRgb> & colors)
+{
+	if(colors.size()> 0)
+	{
+		m_colorsSaved = colors;
+
+		QList<QRgb> callbackColors;
+
+		resizeColorsBuffer(colors.count());
+
+		applyColorModifications(colors, m_colorsBuffer);
+
+		for (int i = 0; i < m_colorsBuffer.count(); i++)
+		{
+			callbackColors.append(qRgb(m_colorsBuffer[i].r>>4, m_colorsBuffer[i].g>>4, m_colorsBuffer[i].b>>4));
+		
+			// Mote
+			m_colorsBuffer[i].r = m_colorsBuffer[i].r >> 4;
+			m_colorsBuffer[i].g = m_colorsBuffer[i].g >> 4;
+			m_colorsBuffer[i].b = m_colorsBuffer[i].b >> 4;
+			PrismatikMath::maxCorrection(254, m_colorsBuffer[i]);
```

Comment:
Sorry I don't understand this.
- What is a mote?
- Why do you shift everything by 4?

Reply:
[it's his particular UDP device and protocol](https://github.com/psieg/Lightpack/issues/297#issuecomment-580920443)

Reply:
- Apologies, that comment should have been removed
- I just copied what the Adalight code was doing in regards to creating the color buffer, is it not required?